portfolio/src/main/java/com/google/sps/data/LongestTaskFirstScheduler.java

```diff
@@ -69,36 +61,147 @@ public SchedulingAlgorithmType getSchedulingAlgorithmType() {
+    }
+
+    return updatedAvailableTimes;
+  }
+
   /**
-   * Tries to find a free time range for a task.
+   * Tries to schedule for a single task by iterating through the currently available time ranges.
+   * Keeps splitting up the task to schedule the currently avaible free time ranges until all of the
+   * task is scheduled across the different time ranges.
    *
-   * @return an Optional that contains the time range scheduled for this task. This Optional object
-   *     is empty if the task cannot be scheduled.
+   * <p>When this method is called, the task may or may not be complete scheduled. The task may be
+   * scheduled into one entire free time range, or the task could be split into multiple time
+   * ranges. When a task is split, new tasks are created to model the segment of the current task
+   * and these new tasks are added to the list of all scheduled tasks.
+   *
+   * @return a list of newly scheduled tasks. If this list is empty, then the current task cannot be
+   *     scheduled.
    */
-  private static Optional<TimeRange> findTimeForTask(
-      Task task, TimeRangeGroup availableTimesGroup) {
-    List<TimeRange> availableTimes = new ArrayList<TimeRange>();
-    Iterator<TimeRange> availableTimesIterator = availableTimesGroup.iterator();
-    Duration taskDuration = task.getDuration();
-    Optional<TimeRange> scheduledTimeRangeOptional = Optional.empty();
+  private List<ScheduledTask> scheduleOneTask(Task task, TimeRangeGroup availableTimesGroup) {
+    List<TimeRange> currentAvailableTimes = constructAvailableTimeRanges(availableTimesGroup);
+    List<ScheduledTask> newScheduledTasks = new ArrayList<ScheduledTask>();
+
+    // If there is no available time ranges anymore,
+    // return the empty list.
+    if (currentAvailableTimes.isEmpty()) {
+      return newScheduledTasks;
+    }
 
-    // Find the first free time range that is longer than the current task's duration.
-    while (availableTimesIterator.hasNext()) {
-      TimeRange currentFreeTimeRange = availableTimesIterator.next();
+    Duration taskDuration = task.getDuration();
+    int taskSegmentCount = 1;
 
-      if (currentFreeTimeRange.duration().compareTo(taskDuration) < 0) {
-        continue;
+    for (TimeRange currentFreeTimeRange : currentAvailableTimes) {
+      if (taskDuration.getSeconds() == 0) {
+        checkForTaskRenaming(newScheduledTasks);
+        return newScheduledTasks;
       }
 
-      // Construct and return the scheduled time range based on the start time of this current time
-      // range.
       Instant scheduledTime = currentFreeTimeRange.start();
-      TimeRange scheduledTimeRange =
-          TimeRange.fromStartEnd(scheduledTime, scheduledTime.plus(taskDuration));
-      scheduledTimeRangeOptional = Optional.of(scheduledTimeRange);
-      break;
+
+      // Reconstruct the task segment's name.
+      String taskName = task.getName() + " (Part " + taskSegmentCount + ")";
```

Comment:
Won't this mean you'll have this "(Part 1)" even if the task fits entirely in one free time range?

EDIT: Oh now I see below how you fix it if the task is scheduled in one time range. I wonder if it's cleaner to instead just move this `taskName` computation inside the `if-else` block. I think the logic can be that if you're inside the `else` block and `taskSegmentCount == ` then you can *not* add the "(Part 1)" suffix, otherwise you do add it?

Reply:
I see! I moved the taskName computation to the else block.

Reply:
I don't think I see this change. Is there another commit you still have to push?

Reply:
I see the change on line 126 now. That seems good thanks.